File: Sorting Algorithms/BUBBLE SORT/bubblesort.py
"""

@DESC : Bubble sort algorithm using Python
@Performance : Worst-case time complexity - O(n ^ 2), Space complexity - O(1)

"""
import time
import random 

class BubbleSort:

    @staticmethod
    def sort(arr, reverse = False):

        for i in range(len(arr)):

            if reverse:
                for j in range(0, len(arr)- i-1):
                    if arr[j] > arr[j+1]:
                        temp = arr[j+1]
                        arr[j+1] = arr[j]
                        arr[j] = temp
            else:
                for j in range(0, len(arr)- i-1):
                    if arr[j] < arr[j+1]:
                        temp = arr[j+1]
                        arr[j+1] = arr[j]
                        arr[j] = temp

        return arr


if __name__ == "__main__":
    algo = BubbleSort()
    arr = [random.randint(1, 100) for _ in range(10000)]
    start = time.time()
    print(algo.sort(arr, reverse = True))
    print("-"*30)
    print(algo.sort(arr, reverse = False))
    end = time.time()
    print("-"*30)
    print(f"Time Elapsed : {end - start}s")

    

File: Sorting Algorithms/BUBBLE SORT/bubblesort.js
// Bubble Sort Algorithm

const bubbleSort = (arr, reverse = false) => {
    let totalSwaps = 0
    for (let i=0 ; i < arr.length ; i++){
      //To compare between adjacent elements
      for (let j=0 ; j < arr.length - i - 1 ; j++){
  
        if (arr[j] > arr[j+1]){
          temp = arr[j+1]
          arr[j+1] = arr[j]
          arr[j] = temp
  
          totalSwaps ++
        }
      }
    }
  
    return [arr, totalSwaps]
  }
  
  
  arr = Array.from( {length: 40}, () => (Math.floor(Math.random()*100)))
  
  start = new Date().getTime()
  const [sorted_arr, totalSwaps] = bubbleSort(arr)
  console.log(sorted_arr)
  console.log(`Total Swaps - ${totalSwaps}`)
  end = new Date().getTime()
  
  console.log(`Time Elapsed - ${end - start} ms`)